internal/utils/gke.go

```diff
@@ -0,0 +1,499 @@
+			Subnetwork: *config.Spec.NetworkConfig.Subnetwork,
+			Network:    *config.Spec.NetworkConfig.Network,
+		}
+	}
+
+	if config.Spec.NetworkPolicy != nil {
+		request.Cluster.NetworkPolicy = &gkeapi.NetworkPolicy{
+			Enabled:  *config.Spec.NetworkPolicy.Enabled,
+			Provider: *config.Spec.NetworkPolicy.Provider,
+		}
+	}
+
+	if config.Spec.PrivateClusterConfig != nil {
+		request.Cluster.PrivateClusterConfig = &gkeapi.PrivateClusterConfig{
+			EnablePrivateEndpoint: *config.Spec.PrivateClusterConfig.EnablePrivateEndpoint,
+			EnablePrivateNodes:    *config.Spec.PrivateClusterConfig.EnablePrivateNodes,
+			MasterIpv4CidrBlock:   config.Spec.PrivateClusterConfig.MasterIpv4CidrBlock,
+			PrivateEndpoint:       config.Spec.PrivateClusterConfig.PrivateEndpoint,
+			PublicEndpoint:        config.Spec.PrivateClusterConfig.PublicEndpoint,
+		}
+	}
+
+	return request, nil
+}
+
+// WaitClusterRemoveExp waits for a cluster to be removed
+func WaitClusterRemoveExp(ctx context.Context, svc *gkeapi.Service, config *gkev1.GKEClusterConfig) (*gkeapi.Operation, error) {
+	var operation *gkeapi.Operation
+	var err error
+
+	for i := 1; i < 12; i++ {
+		time.Sleep(time.Duration(i*i) * time.Second)
+		operation, err = svc.Projects.Locations.Clusters.Delete(ClusterRRN(config.Spec.ProjectID, config.Spec.Region, config.Spec.ClusterName)).Context(ctx).Do()
+		if err == nil {
+			return operation, nil
+		} else if !strings.Contains(err.Error(), "Please wait and try again once it is done") {
+			break
+		}
+	}
+	return operation, err
+}
+
+// GetClientset returns the clientset for a cluster
+func GetClientset(cluster *gkeapi.Cluster, ts oauth2.TokenSource) (kubernetes.Interface, error) {
+	capem, err := base64.StdEncoding.DecodeString(cluster.MasterAuth.ClusterCaCertificate)
+	if err != nil {
+		return nil, err
+	}
+	host := cluster.Endpoint
+	if !strings.HasPrefix(host, "https://") {
+		host = fmt.Sprintf("https://%s", host)
+	}
+
+	config := &rest.Config{
+		Host: host,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAData: capem,
+		},
+		WrapTransport: func(rt http.RoundTripper) http.RoundTripper {
+			return &oauth2.Transport{
+				Source: ts,
+				Base:   rt,
+			}
+		},
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientset, nil
+}
+
+// GetGKEClient returns a gke client capable of making requests on behalf of the services account
+func GetGKEClient(ctx context.Context, secretsCache wranglerv1.SecretCache, config *gkev1.GKEClusterConfig) (*gkeapi.Service, error) {
+	ns, id := ParseCredential(config.Spec.CredentialContent)
+	secret, err := secretsCache.Get(ns, id)
+
+	if err != nil {
+		return nil, err
+	}
+
+	dataBytes := secret.Data["gkeCredentialConfig-data"]
+
+	data := string(dataBytes)
+
+	return GetServiceClient(ctx, data)
+}
+
+// ValidateCreateRequest checks a config for the ability to generate a create request
+func ValidateCreateRequest(config *gkev1.GKEClusterConfig) error {
```

Comment:
On create it would be expected that all nillable fields in the config are not nil. Some fields that are missing from validation are `MasterAuthorizedNetworksConfig` and `privateCusterConfig`. A nil value in kev2 means the field is not "managed" meaning that we will ignore it during sync and update. This applies to maps and slices as well. This does not apply if a cluster is created through gke/rancher as all fields should be managed in this case. This is particularly important for fields at the topmost level of Spec since I believe kubernetes doesn't have a concept of partial updates for structs, slices, and maps. I may have worded this poorly so I can clarify if necessary.

Reply:
Added validation for these two attributes as well as all the others that were missing in 4758fc5fc330fdf6bd4e089f503df43a7c0af916.

There are some attributes in the spec definition that are pointers to structs whose attributes are also pointers, for example 

```
type GKEClusterConfigSpec struct {
<snipped>
    NetworkPolicy                  *NetworkPolicy                  `json:"networkPolicy,omitempty"`
}
<snipped>
type NetworkConfig struct {
     Network    *string `json:"network,omitempty"`
     Subnetwork *string `json:"subnetwork,omitempty"`
}
```

Do you have any sense of whether it's necessary to have both levels be pointers? Could I instead either make NetworkPolicy a regular struct with its attributes as pointers, or the other way around with NetworkPolicy a pointer with its attributes regular strings? `ClusterAddons` and `MasterAuthorizedNetworksConfig` are also inconsistent with one another.

Reply:
Discussed this earlier but to document, the other way around where NetworkPolicy is a point but field are regular should be fine.